crates/uv/src/commands/project/init.rs

```diff
@@ -41,79 +43,185 @@ pub(crate) async fn init(
+
+            match explicit_path {
+                // Initialized a project in the current directory.
+                None => {
+                    writeln!(printer.stderr(), "Initialized project `{}`", name.cyan())?;
+                }
+                // Initialized a project in the given directory.
+                Some(path) => {
+                    let path = std::path::absolute(&path)
+                        .unwrap_or_else(|_| path.simplified().to_path_buf());
+                    writeln!(
+                        printer.stderr(),
+                        "Initialized project `{}` at `{}`",
+                        name.cyan(),
+                        path.display().cyan()
+                    )?;
+                }
+            }
+        }
     }
 
-    // Default to the directory name if a name was not provided.
-    let name = match name {
-        Some(name) => name,
-        None => {
-            let name = path
-                .file_name()
-                .and_then(|path| path.to_str())
-                .context("Missing directory name")?;
+    Ok(ExitStatus::Success)
+}
 
-            PackageName::new(name.to_string())?
+#[allow(clippy::fn_params_excessive_bools)]
+async fn init_script(
+    script_path: &Path,
+    python: Option<String>,
+    connectivity: Connectivity,
+    python_preference: PythonPreference,
+    python_downloads: PythonDownloads,
+    cache: &Cache,
+    printer: Printer,
+    no_workspace: bool,
+    no_readme: bool,
+    no_pin_python: bool,
+    package: bool,
+    native_tls: bool,
+) -> Result<()> {
+    if no_workspace {
+        warn_user_once!("`--no_workspace` is a no-op for Python scripts, which are standalone");
+    }
+    if no_readme {
+        warn_user_once!("`--no_readme` is a no-op for Python scripts, which are standalone");
+    }
+    if package {
+        warn_user_once!("`--package` is a no-op for Python scripts, which are standalone");
+    }
+    let client_builder = BaseClientBuilder::new()
+        .connectivity(connectivity)
+        .native_tls(native_tls);
+
+    let reporter = PythonDownloadReporter::single(printer);
+
+    // If the file already exists, read its content.
+    let content = match fs_err::tokio::read(script_path).await {
+        Ok(metadata) => {
+            // If the file is already a script, raise an error.
+            if ScriptTag::parse(&metadata)?.is_some() {
+                anyhow::bail!(
+                    "`{}` is already a PEP 723 script; use `{}` to execute it",
+                    script_path.simplified_display().cyan(),
+                    "uv run".green()
+                );
+            }
+
+            Some(metadata)
+        }
+        Err(err) if err.kind() == std::io::ErrorKind::NotFound => None,
+        Err(err) => {
+            return Err(anyhow::Error::from(err).context(format!(
+                "Failed to read script at `{}`",
+                script_path.simplified_display().cyan()
+            )));
         }
     };
```

Comment:
I refactored this to do a single I/O operation -- so, we try to read the script, and then we handle appropriate errors. It goes against the idea of using error handling for control flow somewhat, but it avoids race conditions are [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) errors: in the previous implementation, between the time you checked the file metadata and the time you actually read the file, the file could be removed or replaced.

Reply:
This makes sense — this operation felt a little like a "conversational" transaction of sorts, the new implementation is much cleaner